src/rpc/core_rpc_server.cpp

```diff
@@ -1111,9 +1227,71 @@ namespace cryptonote
     return true;
   }
   //------------------------------------------------------------------------------------------------------------------------------
+  bool core_rpc_server::should_use_bootstrap_daemon()
+  {
+    if (m_bootstrap_daemon_address.empty())
+      return false;
+    if (!m_should_use_bootstrap_daemon)
+    {
+      MINFO("The local daemon is fully synced. Not switching back to the bootstrap daemon");
+      return false;
+    }
+    auto current_time = std::chrono::system_clock::now();
+    if (current_time - m_bootstrap_height_check_time > std::chrono::seconds(30))  // update every 30s
+    {
+      m_bootstrap_height_check_time = current_time;
+
+      uint64_t top_height;
+      crypto::hash top_hash;
+      m_core.get_blockchain_top(top_height, top_hash);
+      ++top_height; // turn top block height into blockchain height
+      uint64_t target_height = m_core.get_target_blockchain_height();
+
+      m_should_use_bootstrap_daemon = target_height ? top_height < target_height : true;
```

Comment:
This doesn't seem to be addressed yet. There's a data race on `m_should_use_bootstrap_daemon` (which _currently_ only causes log spam). It might be worth adding a mutex here too, no?

Reply:
Without a mutex, `m_should_use_bootstrap_daemon` could in theory jump from `false` back to `true`, when a dishonest node causes the target height to decrease to 0: https://github.com/monero-project/monero/blob/0b5994fbec79480cd8aa62ec1e10bc7d7fe63efe/src/cryptonote_protocol/cryptonote_protocol_handler.inl#L1721

Reply:
`m_should_use_bootstrap_daemon = target_height == 0 || top_height < target_height;` looks more expressive to me.

Reply:
I don't understand this bit. `m_core.get_target_blockchain_height();` is from your local daemon, so it has nothing to do with any adversary.

Reply:
Yes, it comes from your local node, but it can be influenced by a remote node in such a way that it returns `0`, even if it was set to something bigger before. See the code I linked to in https://github.com/monero-project/monero/pull/3165#discussion_r163431267.

Reply:
Ah, I was missing that point, I think you're right. I've added a mutex lock.

@moneromooo-monero 
Could you confirm?

Reply:
This doesn't seem to check the bootstrap node is actually synced (or at least further synced than the local node).
The comment about the mutex seems correct, if that's what you were asking to confirm.

Reply:
No, it doesn't check the height of the bootstrap node, blindly assuming that the bootstrap node is fully synced. Should it check the height? And what should be the desired behavior when the bootstrap node is not synced?

Reply:
It should at least reject the node if its height is < the current height. If > and we are not synced, I think it's fine to do it as it currently does. AFAICT, the test here uses the bootstrap dameon if we are not synced, which means if we're at about the same sync level as the other node, and are both syncing, we'll still be using the node till it is synced (and if we finish syncing first, we'll still be stuck to the node until it gets synced). That seems wrong.

Moreover, the test (  m_should_use_bootstrap_daemon = target_height ? top_height < target_height : true; ) means that if there is no target height (ie, we're synced from startup), then we're going to use the node. That seems wrong too.

Unless I'm misreading.

Reply:
OK